fe_rtos/src/fe_alloc.rs

```diff
@@ -5,36 +5,33 @@ use crate::task::get_cur_task;
 use alloc::boxed::Box;
 use alloc::vec::Vec;
 use core::alloc::{GlobalAlloc, Layout};
-use core::mem::size_of;
 use core::ptr::null_mut;
 use fe_osi::allocator::LayoutFFI;
 
 /******************************************************************************
-* The structure of a heap memory block is the following:
-*   block header: size_of(usize) bytes, 2*size_of(usize) byte aligned
-*   block data: The data requested by the process
-*   padding: padding required to make the next header block 2*size_of(usize) byte aligned
-*   block footer: identical to the header
+* The structure of a heap is the following:
+*    ---------------------------------------------
+*   | free list |         usable heap            |
+*    ---------------------------------------------
+*   The beginning of the heap is reserved for the free list.
+*   The ith bit of the free list is 0 if the ith block in the usable heap is
+*   unallocated and is 1 otherwise.
+*   The "usable heap" is the part of the heap that will be used to service
+*   dynamic memory allocations.
 *
-* The headers have the following structure:
-*   bits 32-3: The size of the data including padding and headers/footers
-*   bit 2: 1 if the block is allocated. 0 otherwise
-*   bit 1: 1 if the block is the first block. 0 otherwise
-*   bit 0: 1 if the block is the last block. 0 otherwise
+*   In the code below, the usable heap will simple be refered to as the "heap."
 ******************************************************************************/
 
-const HEADER_ALIGN: usize = (2 * size_of::<usize>()) - 1;
-const SIZE_MASK: usize = !HEADER_ALIGN;
-const ALLOC_MASK: usize = 0x4;
-const FIRST_MASK: usize = 0x2;
-const LAST_MASK: usize = 0x1;
+const BLOCK_SIZE: usize = 0x10;
 static ALLOC_LOCK: Spinlock = Spinlock::new();
```

Comment:
It's been a hot minute since I have actually written Rust code, but is it possible to make these global variables fields of the `KernelAllocator` struct, and then instantiate a `KernelAllocator`? I feel like you might be able to take better advantage of Rust's type system that way and might make some of the manual locking a bit less C-like.

Reply:
I'm not opposed to this idea, but I think doing this might be harder than it appears on the surface. First the `GlobalAlloc` `alloc` and `dealloc` methods take `&self` instead of `&mut self` as parameters. That wouldn't be insurmountable, but potentially inconvenient to work around.
The other thing is that if we add the ability for FeRTOS to load and execute arbitrary binaries, those binaries would likely use `fe_osi`'s allocator, which would be separate from the kernel's.
Maybe we could open an issue on this to remind us to look into this in the future, because I also am not a fan of global mutable variables.